File: app-python-zoho/observability/sanitize.py
"""Payload sanitization helpers for structured events."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

MAX_STRING_LENGTH = 2000
MAX_ARRAY_ITEMS = 50
MAX_OBJECT_KEYS = 100
MAX_DEPTH = 4

TRUNCATED_SUFFIX = "...[truncated]"
REDACTED_VALUE = "[redacted]"
SENSITIVE_KEY_FRAGMENTS = (
    "token",
    "secret",
    "password",
    "authorization",
    "cookie",
    "set-cookie",
    "api_key",
    "apikey",
    "access_token",
    "refresh_token",
)


def _truncate_string(value: str) -> tuple[str, bool]:
    text = str(value)
    if len(text) <= MAX_STRING_LENGTH:
        return text, False
    limit = max(0, MAX_STRING_LENGTH - len(TRUNCATED_SUFFIX))
    return f"{text[:limit]}{TRUNCATED_SUFFIX}", True


def _is_sensitive_key(key: Any) -> bool:
    lowered = str(key or "").strip().lower()
    return any(fragment in lowered for fragment in SENSITIVE_KEY_FRAGMENTS)


def _summarize_binary(value: Any) -> str:
    try:
        size = len(value)
    except Exception:
        size = "unknown"
    return f"[binary:{type(value).__name__} size={size}]"


def _summarize_object(value: Any) -> str:
    try:
        text = str(value)
    except Exception:
        text = f"<unprintable {type(value).__name__}>"
    truncated, _ = _truncate_string(text)
    return truncated


def sanitize_value(value: Any, *, max_depth: int = MAX_DEPTH) -> Any:
    sanitized, _ = _sanitize_value(value, depth=0, max_depth=max_depth)
    return sanitized


def _sanitize_value(value: Any, *, depth: int, max_depth: int) -> tuple[Any, bool]:
    try:
        if value is None or isinstance(value, (bool, int, float)):
            return value, False

        if isinstance(value, str):
            return _truncate_string(value)

        if isinstance(value, bytes):
            return _summarize_binary(value), True

        if isinstance(value, BaseException):
            return _truncate_string(f"{type(value).__name__}: {value}")

        if depth >= max_depth:
            return f"[truncated:{type(value).__name__}:depth]", True

        if isinstance(value, Mapping):
            items = list(value.items())
            truncated = len(items) > MAX_OBJECT_KEYS
            result = {}
            for key, item_value in items[:MAX_OBJECT_KEYS]:
                safe_key = str(key)
                if _is_sensitive_key(safe_key):
                    result[safe_key] = REDACTED_VALUE
                    truncated = True
                    continue
                sanitized_item, item_truncated = _sanitize_value(
                    item_value,
                    depth=depth + 1,
                    max_depth=max_depth,
                )
                result[safe_key] = sanitized_item
                truncated = truncated or item_truncated
            if len(items) > MAX_OBJECT_KEYS:
                result["_truncated_keys"] = len(items) - MAX_OBJECT_KEYS
            return result, truncated

        if isinstance(value, (list, tuple, set, frozenset)):
            values = list(value)
            truncated = len(values) > MAX_ARRAY_ITEMS
            result = []
            for item in values[:MAX_ARRAY_ITEMS]:
                sanitized_item, item_truncated = _sanitize_value(
                    item,
                    depth=depth + 1,
                    max_depth=max_depth,
                )
                result.append(sanitized_item)
                truncated = truncated or item_truncated
            if len(values) > MAX_ARRAY_ITEMS:
                result.append(f"[+{len(values) - MAX_ARRAY_ITEMS} more items]")
            return result, truncated

        return _summarize_object(value), True
    except Exception as ex:
        return f"[sanitize_error:{type(ex).__name__}]", True


def sanitize_event(event: dict) -> dict:
    try:
        raw = dict(event or {})
    except Exception:
        raw = {"event": "unknown.event", "component": "unknown"}

    sanitized, truncated = _sanitize_value(raw, depth=0, max_depth=MAX_DEPTH)
    if not isinstance(sanitized, dict):
        sanitized = {"data": sanitized}
        truncated = True
    sanitized["truncated"] = bool(truncated or sanitized.get("truncated"))
    return sanitized

File: autohom-extension/observability/sanitize.js
(function initAutoHomObservabilitySanitize(globalScope) {
  const MAX_STRING_LENGTH = 2000;
  const MAX_ARRAY_ITEMS = 50;
  const MAX_OBJECT_KEYS = 100;
  const MAX_DEPTH = 4;
  const TRUNCATED_SUFFIX = "...[truncated]";
  const REDACTED_VALUE = "[redacted]";
  const SENSITIVE_KEY_FRAGMENTS = [
    "token",
    "secret",
    "password",
    "authorization",
    "cookie",
    "set-cookie",
    "api_key",
    "apikey",
    "access_token",
    "refresh_token",
  ];

  function truncateString(value) {
    const text = String(value);
    if (text.length <= MAX_STRING_LENGTH) {
      return { value: text, truncated: false };
    }
    const limit = Math.max(0, MAX_STRING_LENGTH - TRUNCATED_SUFFIX.length);
    return {
      value: `${text.slice(0, limit)}${TRUNCATED_SUFFIX}`,
      truncated: true,
    };
  }

  function isSensitiveKey(key) {
    const lowered = String(key || "").trim().toLowerCase();
    return SENSITIVE_KEY_FRAGMENTS.some((fragment) => lowered.includes(fragment));
  }

  function summarizeBinary(value) {
    const size = typeof value?.size === "number"
      ? value.size
      : typeof value?.byteLength === "number"
        ? value.byteLength
        : "unknown";
    return `[binary:${value?.constructor?.name || typeof value} size=${size}]`;
  }

  function summarizeObject(value) {
    try {
      return truncateString(String(value)).value;
    } catch (_) {
      return `[object:${value?.constructor?.name || typeof value}]`;
    }
  }

  function sanitizeValue(value, options = {}) {
    const sanitized = sanitizeInternal(value, 0, options.maxDepth ?? MAX_DEPTH);
    return sanitized.value;
  }

  function sanitizeInternal(value, depth, maxDepth) {
    try {
      if (value == null || typeof value === "boolean" || typeof value === "number") {
        return { value, truncated: false };
      }

      if (typeof value === "string") {
        return truncateString(value);
      }

      if (typeof value === "bigint") {
        return { value: String(value), truncated: true };
      }

      if (typeof value === "function") {
        return { value: `[function:${value.name || "anonymous"}]`, truncated: true };
      }

      if (value instanceof Error) {
        return truncateString(`${value.name || "Error"}: ${value.message || ""}`);
      }

      if (
        (typeof Node !== "undefined" && value instanceof Node) ||
        (typeof Element !== "undefined" && value instanceof Element)
      ) {
        return { value: "[dom-node]", truncated: true };
      }

      if (
        value instanceof ArrayBuffer ||
        ArrayBuffer.isView(value) ||
        (typeof Blob !== "undefined" && value instanceof Blob) ||
        (typeof File !== "undefined" && value instanceof File)
      ) {
        return { value: summarizeBinary(value), truncated: true };
      }

      if (depth >= maxDepth) {
        return {
          value: `[truncated:${value?.constructor?.name || typeof value}:depth]`,
          truncated: true,
        };
      }

      if (Array.isArray(value)) {
        let truncated = value.length > MAX_ARRAY_ITEMS;
        const result = [];
        for (const item of value.slice(0, MAX_ARRAY_ITEMS)) {
          const sanitizedItem = sanitizeInternal(item, depth + 1, maxDepth);
          result.push(sanitizedItem.value);
          truncated = truncated || sanitizedItem.truncated;
        }
        if (value.length > MAX_ARRAY_ITEMS) {
          result.push(`[+${value.length - MAX_ARRAY_ITEMS} more items]`);
        }
        return { value: result, truncated };
      }

      if (typeof value === "object") {
        const keys = Object.keys(value);
        let truncated = keys.length > MAX_OBJECT_KEYS;
        const result = {};
        for (const key of keys.slice(0, MAX_OBJECT_KEYS)) {
          if (isSensitiveKey(key)) {
            result[key] = REDACTED_VALUE;
            truncated = true;
            continue;
          }
          const sanitizedItem = sanitizeInternal(value[key], depth + 1, maxDepth);
          result[key] = sanitizedItem.value;
          truncated = truncated || sanitizedItem.truncated;
        }
        if (keys.length > MAX_OBJECT_KEYS) {
          result._truncated_keys = keys.length - MAX_OBJECT_KEYS;
        }
        return { value: result, truncated };
      }

      return { value: summarizeObject(value), truncated: true };
    } catch (error) {
      return {
        value: `[sanitize_error:${error?.name || "Error"}]`,
        truncated: true,
      };
    }
  }

  function sanitizeEvent(event) {
    const baseEvent = event && typeof event === "object" ? { ...event } : {
      component: "unknown",
      event: "unknown.event",
    };
    const sanitized = sanitizeInternal(baseEvent, 0, MAX_DEPTH);
    const result = sanitized.value && typeof sanitized.value === "object"
      ? sanitized.value
      : { data: sanitized.value };
    result.truncated = Boolean(sanitized.truncated || result.truncated);
    return result;
  }

  globalScope.AutoHomObservabilitySanitize = {
    MAX_STRING_LENGTH,
    MAX_ARRAY_ITEMS,
    MAX_OBJECT_KEYS,
    MAX_DEPTH,
    sanitizeValue,
    sanitizeEvent,
  };
})(globalThis);
